sparse/mlir_backend/tests/test_simple.py

```diff
@@ -108,93 +123,92 @@ def test_dense_format(dtype, shape):
     np.testing.assert_equal(actual, data)
 
 
+def assert_array_equal(
+    expected: sparse.Array,
+    actual: sparse.Array,
+    /,
+    *,
+    same_format: bool = True,
+    same_dtype: bool = True,
+    data_test_fn: typing.Callable[[np.ndarray, np.ndarray], None] = np.testing.assert_array_equal,
+) -> None:
+    if same_format:
+        assert expected.format == actual.format
+
+    if same_dtype:
+        assert expected.dtype == actual.dtype
+
+    assert expected.shape == actual.shape
+    actual = actual.asformat(expected.format)
+
+    carrs_expected = expected.get_constituent_arrays()
+    carrs_actual = actual.get_constituent_arrays()
+
+    for e, a in zip(carrs_expected[:-1], carrs_actual[:-1], strict=True):
+        assert e.dtype == a.dtype
+        np.testing.assert_equal(e, a)
+
+    data_test_fn(carrs_expected[-1], carrs_actual[-1])
+
+
 @parametrize_dtypes
-def test_2d_constructors(rng, dtype):
+@parametrize_scipy_fmt
+def test_roundtrip(rng, dtype, format):
     SHAPE = (80, 100)
     DENSITY = 0.6
     sampler = generate_sampler(dtype, rng)
-    csr = sps.random_array(SHAPE, density=DENSITY, format="csr", dtype=dtype, random_state=rng, data_sampler=sampler)
-    csc = sps.random_array(SHAPE, density=DENSITY, format="csc", dtype=dtype, random_state=rng, data_sampler=sampler)
-    dense = np.arange(math.prod(SHAPE), dtype=dtype).reshape(SHAPE)
-    coo = sps.random_array(SHAPE, density=DENSITY, format="coo", dtype=dtype, random_state=rng, data_sampler=sampler)
-    coo.sum_duplicates()
+    sps_arr = sps.random_array(
+        SHAPE, density=DENSITY, format=format, dtype=dtype, random_state=rng, data_sampler=sampler
+    )
 
-    csr_tensor = sparse.asarray(csr)
-    csc_tensor = sparse.asarray(csc)
-    dense_tensor = sparse.asarray(dense)
-    coo_tensor = sparse.asarray(coo)
-    dense_2_tensor = sparse.asarray(np.arange(100, dtype=dtype).reshape((25, 4)) + 10)
+    sp_arr = sparse.asarray(sps_arr)
+    sps_roundtripped = sparse.to_scipy(sp_arr)
+    assert_sps_equal(sps_arr, sps_roundtripped)
 
-    csr_retured = sparse.to_scipy(csr_tensor)
-    assert_sps_equal(csr_retured, csr)
+    sp_arr_roundtripped = sparse.asarray(sps_roundtripped)
 
-    csc_retured = sparse.to_scipy(csc_tensor)
-    assert_sps_equal(csc_retured, csc)
+    assert_array_equal(sp_arr, sp_arr_roundtripped)
 
-    dense_returned = sparse.to_numpy(dense_tensor)
-    np.testing.assert_equal(dense_returned, dense)
 
-    coo_returned = sparse.to_scipy(coo_tensor)
-    np.testing.assert_equal(coo_returned.todense(), coo.todense())
+@parametrize_dtypes
+def test_roundtrip_dense(rng, dtype):
+    SHAPE = (80, 100)
```

Comment:
For dense let's also parametrize shape, and consider 1D and 3D.

Reply:
Done in https://github.com/pydata/sparse/pull/808/commits/406330b082d9a926095874e588c5db630394acaf.